File: source/downloads/code/LeetCode/Java/Reorder-List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void reorderList(ListNode head) {
        if(head==null || head.next==null) return ;
        
        // Find middle
        ListNode prev = new ListNode(-1);
        prev.next = head;
        ListNode p1 = prev;
        ListNode p2 = prev;
        while(p2!=null && p2.next!=null){
            p1 = p1.next;
            p2 = p2.next.next;
        }

        ListNode right = p1.next;
        ListNode left = head;
        p1.next = null;

        // Reverse the right list

        ListNode prevNode = right;
        ListNode nextNode = right.next;
        right.next = null;

        while(nextNode!=null){
            ListNode tmp = nextNode.next;
            nextNode.next = prevNode;
            prevNode = nextNode;
            nextNode = tmp;
        }
        right = prevNode;

        // Merge left and right
        while(right!=null){
            ListNode leftNext = left.next;
            ListNode rightNext = right.next;
            left.next = right;
            right.next = leftNext;
            left = leftNext;
            right = rightNext;
        
        }

    }
}

File: downloads/code/LeetCode/Python/Reorder-List.py
#Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: void Do not return anything, modify head in-place instead.
        """
        if head == None or head.next == None:
            return 
        # split the list
        prevNode = ListNode(-1)
        prevNode.next = head
        slow = prevNode
        fast = prevNode

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        right = slow.next
        slow.next = None
        left = head

        # reverse the right
        prev = right
        nextNode = right.next
        prev.next = None
        while nextNode!=None:
            tmp = nextNode.next
            nextNode.next = prev
            prev = nextNode
            nextNode = tmp

        right = prev

        # Merge the list
        while right != None:
            leftNext = left.next
            rightNext = right.next
            left.next = right
            right.next = leftNext
            left = leftNext
            right = rightNext

